Packages/com.unity.inputsystem/InputSystem/TypeSafeAPI/Input.actions.cs

```diff
@@ -0,0 +1,90 @@
+    {
+        /// <summary>
+        /// Enables access to the underlying Input Action for advanced functionality such as rebinding.
+        /// </summary>
+        public InputAction action => m_Action;
+
+        /// <see cref="InputAction.IsPressed"/>
+        public bool isPressed => m_Action.IsPressed();
+        /// <see cref="InputAction.WasPressedThisFrame"/>
+        public bool wasPressedThisFrame => m_Action.WasPressedThisFrame();
+        /// <see cref="InputAction.WasReleasedThisFrame"/>
+        public bool wasReleasedThisFrame => m_Action.WasReleasedThisFrame();
+
+        /// <summary>
+        /// Returns the current value of the Input Action.
+        /// </summary>
+        public TActionType value
+        {
+            get
+            {
+                try
+                {
+                    // it should be unusual for ReadValue to throw because in most cases instances of this class
+                    // will be created through the source generator, and that can catch mismatched control types
+                    // at compile time and throw compiler errors, but it will always be possible to dynamically
+                    // add incompatible bindings, and the best we can do then is to catch the exceptions
+                    // thrown when we try to read from those controls.
+                    return m_Action.ReadValue<TActionType>();
+                }
+                catch (InvalidOperationException ex)
+                {
+                    Debug.LogWarning(ex.Message);
+                }
+
+                return m_Action.ReadDefaultValue<TActionType>(
+                    new BindingIndex(m_Action, m_Action.activeBindingIndex, BindingIndex.IndexType.IncludeCompositeParts)
+                        .ToIndexWithoutCompositeParts().value);
+            }
+        }
+
+        /// <summary>
+        /// A high-level collection of Input Bindings that by default skips over the composite part bindings
+        /// for purposes of indexing.
+        /// </summary>
+        /// <remarks>
+        ///
+        /// </remarks>
+        public InputBindingsCollection bindings =>
+            new InputBindingsCollection(m_Action, InputBindingsCollection.EnumerationBehaviour.SkipCompositeParts);
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="action"></param>
+        /// <remarks>
+        /// </remarks>
+        public Input(InputAction action)
+        {
+            Debug.Assert(action != null);
+
+            m_Action = action ?? throw new ArgumentNullException(nameof(action));
+            m_Action.Enable();
+        }
+
+        /// <summary>
+        /// True when the action is pressed.
+        /// </summary>
+        /// <param name="input"></param>
+        /// <remarks>
+        /// </remarks>
+        public static implicit operator bool(Input<TActionType> input)
+        {
+            return input.m_Action.IsPressed();
+        }
```

Comment:
I don't know if we should add this implicit operator, especially since it is using `IsPressed` instead of something like `phase == InputActionPhase.Performed`. If you have a Button input action with a Hold interaction on the binding, `IsPressed` will be `true` the moment you press the button, but it won't be performed until 0.4 seconds later in the default. Using `IsPressed` here seems bad for the default.

Since there's subtle differences between `IsPressed`, `WasPerformedThisFrame`, and `phase == InputActionPhase.Performed`, I think it should remain explicit for what the user wants, and I think this operator should be deleted.

Reply:
Thanks for looking this over @chris-massie. Good points, I think this should be okay to remove entirely.